localstack/services/s3/provider.py

```diff
@@ -802,7 +804,28 @@ def upload_part(self, context: RequestContext, request: UploadPartRequest) -> Up
                 ArgumentValue=part_number,
             )
 
-        part = body.read() if (body := request.get("Body")) else b""
+        body = request.get("Body")
+        headers = context.request.headers
+        if body and (
+            (content_sha_256 := headers.get("x-amz-content-sha256") or "")
```

Comment:
According to the docs you linked, the AWS Chunked Encoding also has to be declared in the `Content-Encoding` header. This might be a cleaner way to define that the payload is encoded with AWS' proprietary chunk encoding.

Reply:
I've just found the original issue we've had, and this was the sent request:

```
PUT /bucket1/data HTTP/1.1
Host: 127.0.0.1:14572
User-Agent: MinIO (linux; amd64) minio-go/v6.0.35
Content-Length: 185
Authorization: AWS4-HMAC-SHA256 Credential=foo/20190918/us-east-1/s3/aws4_request,SignedHeaders=host;x-amz-content-sha256;x-amz-date;x-amz-decoded-content-length,Signature=9ff757b67570b0d555fed0e4944ce012ebcb3f963b40044296f3fb89cdb739d8
Content-Type: audio/mpeg
X-Amz-Content-Sha256: STREAMING-AWS4-HMAC-SHA256-PAYLOAD
X-Amz-Date: 20190918T051509Z
X-Amz-Decoded-Content-Length: 13

d;chunk-signature=af5e6c0a698b0192e9aa5d9083553d4d241d81f69ec62b184d05c509ad5166af
Hello  



0;chunk-signature=f2a50a8c0ad4d212b579c2489c6d122db88d8a0d0b987ea1f3e9d081074a5937
```

I'm worried the `Content-Encoding` header is not always respected...

Reply:
And this request would work against AWS? It seems like a test-request coming from a minio client? The spec quite explicitly says that you have to set the content encoding: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html

Reply:
I'm not sure, because I would need to properly sign it and I can't do it with boto. I'll need a test setup with Java or .NET and manually change the headers to see if AWS will reject it. I guess it should.
From another issue (https://github.com/localstack/localstack/issues/6659), the `Content-Encoding` is properly set from the Java client. I think we can follow the spec like you indicated, and if issues arise because AWS clients do not follow the specs then we will fix it.

Reply:
Found the [minio implementation](https://github.com/minio/minio-go/blob/b75cfd43c196482756ff95460c779793e18d3ce4/pkg/signer/request-signature-v4.go#L292) now, it seems they are indeed setting the header, so we should be good, it was an old issue.